gpu.hpp

```diff
@@ -794,6 +809,212 @@ inline void check(bool condition, const char *message,
   }
 }
 
+/**
+ * @brief Pumps events until the provided future is ready.
+ *
+ * This helper template function continuously checks the status of the provided
+ * std::future<T> until it becomes ready. On Emscripten builds, it yields
+ * control to the JavaScript event loop using emscripten_sleep to allow
+ * asynchronous callbacks to execute. On other platforms, it processes events
+ * from the given WGPUInstance using wgpuInstanceProcessEvents. Once the future
+ * is ready, its value is returned.
+ *
+ * @tparam T The type of the value contained in the future.
+ * @param instance The WGPUInstance used to process events.
+ * @param f The future to wait on.
+ * @return T The value retrieved from the ready future.
+ *
+ * @code
+ * std::future<WGPUDevice> deviceFuture = requestDeviceAsync(adapter,
+ * devDescriptor); WGPUDevice device = wait(instance, deviceFuture);
+ * @endcode
+ */
+template <typename T> T wait(Context &ctx, std::future<T> &f) {
+#ifdef __EMSCRIPTEN__
+  // Poll until the future is ready.
+  while (f.wait_for(std::chrono::milliseconds(0)) !=
+         std::future_status::ready) {
+    // Yield control to the JS event loop.
+    emscripten_sleep(1);
+  }
+  return f.get();
+#else
+  while (f.wait_for(std::chrono::milliseconds(0)) !=
+         std::future_status::ready) {
+    wgpuInstanceProcessEvents(ctx.instance);
+  }
+  return f.get();
+#endif
+}
+
+// Context Callbacks & Helpers
+
+/**
+ * @brief Waits for the provided std::future<T> to become ready by polling its
+ * status.
+ *
+ * This helper template function continuously checks the status of the provided
+ * std::future<T> until it is ready. On Emscripten builds, it yields control to
+ * the JavaScript event loop using emscripten_sleep(1) for smooth asynchronous
+ * behavior. On non-Emscripten platforms, it sleeps for a short duration (10
+ * milliseconds) between checks. Once the future is ready, its value is
+ * returned.
+ *
+ * @tparam T The type of the value contained in the future.
+ * @param f The future to wait on.
+ * @return T The value retrieved from the ready future.
+ *
+ * @code
+ * std::future<Context> contextFuture = createContext();
+ * Context ctx = waitForContextFuture(contextFuture);
+ * @endcode
+ */
+template <typename T> T waitForContextFuture(std::future<T> &f) {
+#ifdef __EMSCRIPTEN__
+  while (f.wait_for(std::chrono::milliseconds(0)) !=
+         std::future_status::ready) {
+    emscripten_sleep(1); // Yield back to the JS event loop.
+  }
+  return f.get();
+#else
+  while (f.wait_for(std::chrono::milliseconds(0)) !=
+         std::future_status::ready) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
```

Comment:
What if we set this as a function argument? Different applications might have different needs /(we could have an overload where 10ms is a default even).